File: android/android/src/de/benibela/videlibri/notifications/Notifier.java
package de.benibela.videlibri.notifications;

import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.support.v4.app.NotificationCompat;

import de.benibela.videlibri.Accounts;
import de.benibela.videlibri.LendingList;
import de.benibela.videlibri.R;
import de.benibela.videlibri.Util;
import de.benibela.videlibri.VideLibriApp;
import de.benibela.videlibri.jni.Bridge;

import static android.content.Context.NOTIFICATION_SERVICE;

public class Notifier {
    static private String lastNotificationTitle = "";
    static private String lastNotificationText = "";
    static private long lastNotificationTime = 0;
    static private final String DUEDATE_CHANNEL = "duedate";

    static private String [] getNotifications(Context context){
        if (Accounts.INSTANCE.size() == 0) return new String[] {
                Util.tr(context, R.string.notificationNoAccountsTitle),
                Util.tr(context, R.string.notificationNoAccounts)
        };

        String [] notification = Bridge.VLGetNotifications();

        if (notification == null) return null;

        return notification;
    }

    static private final int NOTIFICATION_ID = 8239238;
    static private void cancelNotification(Context context) {
        NotificationManager nm = ((NotificationManager)context.getSystemService(NOTIFICATION_SERVICE));
        if (nm == null) return;
        nm.cancel(NOTIFICATION_ID);
        lastNotificationTime = 0;
    }
    static private void initChannels(Context context){
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O)
            return;
        NotificationManager nm = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        if (nm == null) return;
        NotificationChannel channel = new NotificationChannel( DUEDATE_CHANNEL, context.getString(R.string.notifications_duedate_channel_name), NotificationManager.IMPORTANCE_DEFAULT);
        channel.setDescription(context.getString(R.string.notifications_duedate_channel_description));
        channel.setShowBadge(true);
        nm.createNotificationChannel(channel);
    }

    /**
     * Show a notification while this service is running.
     * (partly from http://stackoverflow.com/questions/13902115/how-to-create-a-notification-with-notificationcompat-builder)
     */
    static public void updateNotification(Context context) {
        if (context == null) context = VideLibriApp.currentContext();
        if (context == null) return;

        String[] notification = getNotifications(context);

        if (notification == null || notification.length < 2) {
            cancelNotification(context);
            return;
        }

        if (lastNotificationTitle.equals(notification[0]) && lastNotificationText.equals(notification[1])
                && System.currentTimeMillis() - lastNotificationTime < 1000*60*60*23) return;

        initChannels(context);

        lastNotificationTitle = notification[0];
        lastNotificationText = notification[1];
        lastNotificationTime = System.currentTimeMillis();

        NotificationCompat.Builder mBuilder =
                new NotificationCompat.Builder(context, DUEDATE_CHANNEL)
                        .setSmallIcon(VideLibriApp.getMainIcon())
                        .setContentTitle(notification[0])
                        .setContentText(notification[1])
                        .setContentIntent(PendingIntent.getActivity(context, 0, new Intent(context, LendingList.class), 0))
                        .setAutoCancel(true)
                ;

        // Set the info for the views that show in the notification panel.
        //notification.setLatestEventInfo(this, getText(R.string.local_service_label),text, contentIntent);

        // Send the notification.
        NotificationManager nm = ((NotificationManager)context.getSystemService(NOTIFICATION_SERVICE));
        if (nm == null) return;
        nm.notify(NOTIFICATION_ID, mBuilder.build());
    }

}

File: android/android/src/de/benibela/videlibri/notifications/Notifier.kt
package de.benibela.videlibri.notifications

import android.Manifest
import android.app.Activity
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.app.ActivityCompat
import androidx.core.app.NotificationCompat
import androidx.core.content.ContextCompat
import de.benibela.videlibri.Accounts
import de.benibela.videlibri.R
import de.benibela.videlibri.VideLibriApp
import de.benibela.videlibri.activities.LendingList
import de.benibela.videlibri.isReal
import de.benibela.videlibri.jni.Bridge
import de.benibela.videlibri.jni.globalOptionsAndroid
import de.benibela.videlibri.jni.globalOptionsShared
import de.benibela.videlibri.jni.save
import de.benibela.videlibri.utils.notificationManager

object Notifier {
    private const val DUEDATE_CHANNEL = "duedate"
    private const val NOTIFICATION_ID = 8239238
    private const val OUTDATED_DATA_PERIOD_DAYS: Long = 7

    private fun skipableNotification(title: String, text: String): Boolean {
        val checkSkipPeriodic: Long = kotlin.math.min(1000L * 60 * 60 * 23, NotificationScheduling.DAILY_CHECK_PERIOD)
        val now = System.currentTimeMillis()
        val notifications = globalOptionsAndroid.notifications
        val canSkip = now - notifications.lastTime < checkSkipPeriodic &&
                      now >= notifications.lastTime &&
                      title == notifications.lastTitle &&
                      text == notifications.lastText

        if (!canSkip) {
            notifications.lastTime = now
            notifications.lastTitle = title
            notifications.lastText = text
            globalOptionsAndroid.save()
        }
        return canSkip
    }


    private fun getNotifications(context: Context): Array<String>? =
            if (Accounts.isEmpty())
                with(context) {
                    if (globalOptionsAndroid.hasBeenStartedAtLeastOnce && globalOptionsAndroid.accountCountBackup <= 0) null
                    else arrayOf(getString(R.string.notificationNoAccountsTitle), getString(R.string.notificationNoAccounts))
                }
            else {
                val n = Bridge.VLGetNotifications()
                if (n?.isNotEmpty() == true) n
                else {
                    Accounts.refreshAccounts()
                    if (Accounts.any { it.isReal && it.lastCheckDate > 0 && it.lastCheckDate <= Bridge.currentPascalDate - OUTDATED_DATA_PERIOD_DAYS })
                        with(context) { arrayOf(getString(R.string.notificationOutdatedDataTitle), getString(R.string.notificationOutdatedData)) }
                    else
                        null
                }
            }

    private fun cancelNotification(context: Context) {
        context.notificationManager?.cancel(NOTIFICATION_ID)
        skipableNotification("", "")
    }

    fun initChannels(context: Context) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O)
            return
        val channel = NotificationChannel(DUEDATE_CHANNEL, context.getString(R.string.notifications_duedate_channel_name), NotificationManager.IMPORTANCE_DEFAULT)
        channel.description = context.getString(R.string.notifications_duedate_channel_description)
        channel.setShowBadge(true)
        context.notificationManager?.createNotificationChannel(channel)
    }


    /**
     * Show a notification while this service is running.
     * (partly from http://stackoverflow.com/questions/13902115/how-to-create-a-notification-with-notificationcompat-builder)
     */
    @JvmStatic fun updateNotification(activityContext: Context?) {
        val context = activityContext ?: VideLibriApp.currentContext() ?: return

        val notification = getNotifications(context) ?: arrayOf()

        if (notification.size < 2) {
            cancelNotification(context)
            return
        }

        if (skipableNotification(notification[0], notification[1]))
            return

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N &&
                context.notificationManager?.areNotificationsEnabled() == false)
            return

        initChannels(context)

        val pendingIntentFlags =
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) PendingIntent.FLAG_IMMUTABLE else 0

        val builder = NotificationCompat.Builder(context, DUEDATE_CHANNEL)
                .setSmallIcon(VideLibriApp.mainIcon)
                .setContentTitle(notification[0])
                .setContentText(notification[1])
                .setContentIntent(PendingIntent.getActivity(context, 0, Intent(context, LendingList::class.java), pendingIntentFlags))
                .setAutoCancel(true)

        // Set the info for the views that show in the notification panel.
        //notification.setLatestEventInfo(this, getText(R.string.local_service_label),text, contentIntent);

        // Send the notification.
        context.notificationManager?.notify(NOTIFICATION_ID, builder.build())
    }

}


const val PERMISSION_REQUEST_CODE_NOTIFICATIONS = 789
fun checkForRequiredNotificationPermission(context: Activity ){
   with (context) {
       if (!globalOptionsAndroid.notifications.enabled) return
       if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU){
           if (ContextCompat.checkSelfPermission(this, Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED) {
               globalOptionsAndroid.notifications.lastAskedForPermission = Bridge.currentPascalDate
               globalOptionsAndroid.save()
               ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.POST_NOTIFICATIONS), PERMISSION_REQUEST_CODE_NOTIFICATIONS)
           }
       }
       Notifier.initChannels(context)
   }
}